src/main/java/com/klodnicki/Bike/v2/service/BikeService.java

```diff
@@ -1,7 +1,126 @@
+
 @Service
-public class BikeService {
+//public class BikeService implements GenericService<BikeForAdminResponseDTO, Bike> {
+//zamienilismy BikeService dependency wstrzykiwane w controllerze na GenericBikeService
+//Żeby móc podstawić różne implementacje BikeService, nie tylko ten jeden konkretny BikeService. Jak zrobisz klasę
+//        BetterBikeService możesz bez żadnych zmian w kontrolerze z niego korzystać od razu
+public class BikeService implements GenericBikeService {
+
+    private final BikeRepository bikeRepository;
+
+    ModelMapper modelMapper = new ModelMapper();
+
+    public BikeService(BikeRepository bikeRepository) {
+        this.bikeRepository = bikeRepository;
+    }
+
+    @Override
+    public BikeForAdminResponseDTO add(BikeRequestDTO bikeRequestDTO) {
+        //using ModelMapper to convert(map) bikeRequestDTO into bike. I need Bike to be able to save it in repo
+        Bike bike = modelMapper.map(bikeRequestDTO, Bike.class);
+        Bike savedBike = bikeRepository.save(bike);
+
+        return convertBikeIntoBikeForAdminResponseDTO(savedBike); // gdyby tutaj brał bike (a nie savedBike) to by klasa
+        //abstrakcyjna nie musiała mieć adnotacji @MappedSuperclass
+    }
+
+    @Override
+    public BikeForAdminResponseDTO findById(Long id) {
+        Bike savedBike = bikeRepository.findById(id).orElseThrow(IllegalArgumentException::new); //TODO: add custom exception
+
+        return convertBikeIntoBikeForAdminResponseDTO(savedBike);
+    }
+
+    @Override
+    public List<BikeForAdminResponseDTO> findAll() {
+        List<BikeForAdminResponseDTO> bikeListDto = new ArrayList<>();
+        Iterable<Bike> bikeList = bikeRepository.findAll();
+
+        for (Bike bike : bikeList) {
+
+            Long id = bike.getId();
+            String serialNumber = bike.getSerialNumber();
+            boolean isRented = bike.isRented();
+            BikeType type = bike.getBikeType();
+            LocalDateTime rentalStartTime = bike.getRentalStartTime();
+            LocalDateTime rentalEndTime = bike.getRentalEndTime();
+            double toBePaid = bike.getAmountToBePaid();
+            GpsCoordinates gps = bike.getGpsCoordinates();
+
+            BikeForAdminResponseDTO bikeDto = new BikeForAdminResponseDTO(id, serialNumber, isRented, type, rentalStartTime,
+                    rentalEndTime, toBePaid, gps);
+            bikeListDto.add(bikeDto);
+        }
+
+        return bikeListDto;
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        bikeRepository.deleteById(id);
+    }
+
+    @Override
+    public BikeForAdminResponseDTO update(Long id, BikeRequestDTO updatedBikeRequestDTO) {
+
+        Bike bike = bikeRepository.findById(id).orElseThrow(IllegalAccessError::new);
+
+        if (updatedBikeRequestDTO.getSerialNumber() != null) {
+            bike.setSerialNumber(updatedBikeRequestDTO.getSerialNumber());
+        }
+
+        bike.setRented(updatedBikeRequestDTO.isRented());
```

Comment:
This line of code for setting rented status is hidden among the ladder of ifs and is easy to miss!
Improve the readability of this whole block, please, by adding a private method where you can keep those if statements.

When you extract those ifs to private method there is also a way to avoid using so many ifs. Hint: all those values you're checking if they are null (with one exception).
You can find a better, more readable and maintainable solution.

Reply:
sure! One way to make this code more readable and maintainable is to use the Optional class in Java. This allows me to perform an action if a value is not null, without using an explicit if statement. 
However  we still need an if statement for getAmountToBePaid() because we’re checking if it’s not zero, not if it’s not null